sdk/search/Azure.Search.Documents/src/Batching/SearchIndexingBufferedSender{T}.cs

```diff
@@ -0,0 +1,606 @@
+        void IDisposable.Dispose() =>
+            DisposeAsync(async: false).EnsureCompleted();
+        #pragma warning restore CA1816 // Dispose methods should call SuppressFinalize
+
+        /// <summary>
+        /// Flush any remaining work and clean up resources.
+        /// </summary>
+        /// <returns>
+        /// A task that will complete when the object has finished disposing.
+        /// </returns>
+        async ValueTask IAsyncDisposable.DisposeAsync() =>
+            await DisposeAsync(async: true).ConfigureAwait(false);
+
+        /// <summary>
+        /// Dispose the sender and flush any remaining indexing actions that
+        /// haven't been sent yet.  This will block until everything's been
+        /// sent.
+        /// </summary>
+        /// <param name="async">Whether to call this sync or async.</param>
+        /// <returns>A Task that will wait until we're disposed.</returns>
+        internal async Task DisposeAsync(bool async)
+        {
+            if (Interlocked.CompareExchange(ref _disposed, 1, 0) == 0)
+            {
+                await _publisher.DisposeAsync(async).ConfigureAwait(false);
+                _keyFieldBarrier.Dispose();
+            }
+        }
+
+        /// <summary>
+        /// Ensure the sender was properly disposed.
+        /// </summary>
+        ~SearchIndexingBufferedSender()
+        {
+            if (_publisher?.IndexingActionsCount > 0)
+            {
+                try
+                {
+                    #pragma warning disable CA1065 // Do not raise exceptions in unexpected locations
+                    throw new InvalidOperationException(
+                        $"{nameof(SearchIndexingBufferedSender<T>)} has {_publisher.IndexingActionsCount} unsent indexing actions.");
+                    #pragma warning restore CA1065 // Do not raise exceptions in unexpected locations
+                }
+                catch (InvalidOperationException)
+                {
+                }
+            }
+        }
+
+        /// <summary>
+        /// Ensure nobody's trying to use this after it's been disposed.
+        /// </summary>
+        private void EnsureNotDisposed()
+        {
+            if (_disposed > 0)
+            {
+                throw new ObjectDisposedException(nameof(SearchIndexingBufferedSender<T>));
+            }
+        }
+        #endregion Dispose
+
+        #region Get Key Field
+        /// <summary>
+        /// Ensure we have a valid KeyFieldAccessor (and probe the type or
+        /// service to find one).
+        /// </summary>
+        /// <param name="async">Whether to run sync or async.</param>
+        /// <param name="cancellationToken">The Cancellation token.</param>
+        /// <returns>A task that will complete after we've checked.</returns>
+        private async Task EnsureKeyFieldAccessorAsync(bool async, CancellationToken cancellationToken)
```

Comment:
Who calls this?

Reply:
AddIndexingActionsInternal needs to.  I yanked out Publisher<T> into its own thing and this got lost in that shuffle.